ligo/src/token/fa2.mligo

```diff
@@ -0,0 +1,176 @@
+// SPDX-FileCopyrightText: 2020 TQ Tezos
+// SPDX-License-Identifier: LicenseRef-MIT-TQ
+
+// Corresponds to Token/FA2.hs module
+
+#include "../types.mligo"
+
+let frozen_token_id: nat = 0n
+let unfrozen_token_id: nat = 1n
+// let (|>) x f = f x
+
+// -----------------------------------------------------------------
+// Helper
+// -----------------------------------------------------------------
+
+let ensure_not_migrated (storage : storage): bool =
+  match storage.migration_status with
+      Not_in_migration -> true
+    | MigratingTo (p) -> true
+    | MigratedTo (p) -> (failwith ("MIGRATED") : bool)
+
+let check_sender (from_ , store : address * storage): bool =
+  if (sender = from_) then true
+  else
+    let key: operator = { owner = from_; operator = sender} in
+    if Big_map.mem key store.operators then
+      true
+    else
+      (failwith("FA2_NOT_OPERATOR"): bool)
+
+let debit_from (amt, from_, token_id : nat * address * token_id) (store : storage): storage =
+  let bal_o = Big_map.find_opt (from_, token_id) store.ledger in
+  match bal_o with
+    Some bal ->
+      let new_bal_int = bal - amt in
+      let updated_store =
+        match Michelson.is_nat (new_bal_int) with
+          Some new_bal ->
+            { store with
+              ledger = Big_map.update (from_, token_id) (Some new_bal) store.ledger
+            }
+        | None ->
+            // TODO: How to fail with some value
+            // (failwith ("FA2_INSUFFICIENT_BALANCE", (amt, bal)) : storage)
+            (failwith ("FA2_INSUFFICIENT_BALANCE") : storage)
+      in updated_store
+  | None ->
+      if (amt = 0n) then store // We allow 0 transfer
+      else (failwith ("FA2_INSUFFICIENT_BALANCE") : storage)
+
+let credit_to (amt, to_, token_id : nat * address * nat) (store : storage): storage =
+  let new_bal =
+    match Big_map.find_opt (to_, token_id) store.ledger with
+      Some bal -> bal + amt
+    | None -> amt
+  in  { store with
+        ledger = Big_map.update (to_, token_id) (Some new_bal) store.ledger
+      }
+
+
+// -----------------------------------------------------------------
+// Transfer
+// -----------------------------------------------------------------
+
+// TODO: how to add to 'where'
```

Comment:
I don't understand this comment: can you explain?

Reply:
@rinn7e still wondering what this comment is about :slightly_smiling_face:

Reply:
I was wondering how to add a `where` clause like in haskell. I guess we can just put the function in `let` as well.

Reply:
I see, can you please remove this comment in this case?